core/drivers/firewall/stm32_risaf.c

```diff
@@ -0,0 +1,730 @@
+			if (io_read32(base + _RISAF_IADDR1))
+				IMSG("Dual port faulty address (IADDR1): %#"PRIx32,
+				     io_read32(base + _RISAF_IADDR1));
+		}
+
+		IMSG("=====================================================\n");
+	};
+}
+
+static __maybe_unused
+bool risaf_is_hw_encryption_enabled(struct stm32_risaf_instance *risaf)
+{
+	return (io_read32(risaf_base(risaf) + _RISAF_SR) &
+		_RISAF_SR_ENCDIS) != _RISAF_SR_ENCDIS;
+}
+
+static TEE_Result
+risaf_check_region_boundaries(struct stm32_risaf_instance *risaf,
+			      struct stm32_risaf_region *region)
+{
+	if (!core_is_buffer_inside(region->addr, region->len,
+				   risaf->pdata.mem_base,
+				   risaf->pdata.mem_size)) {
+		EMSG("Region %#"PRIxPA"..%#"PRIxPA" outside RISAF area %#"PRIxPA"...%#"PRIxPA,
+		     region->addr, region->addr + region->len - 1,
+		     risaf->pdata.mem_base,
+		     risaf->pdata.mem_base + risaf->pdata.mem_size - 1);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!risaf->ddata->granularity ||
+	    (region->addr % risaf->ddata->granularity) ||
+	    (region->len % risaf->ddata->granularity)) {
+		EMSG("RISAF %#"PRIxPA": start/end address granularity not respected",
+		     risaf->pdata.base.pa);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+risaf_check_overlap(struct stm32_risaf_instance *risaf __maybe_unused,
+		    struct stm32_risaf_region *region, unsigned int index)
+{
+	unsigned int i = 0;
+
+	for (i = 0; i < index; i++) {
+		/* Skip region if there's no configuration */
+		if (!region[i].cfg)
+			continue;
+
+		if (core_is_buffer_intersect(region[index].addr,
+					     region[index].len,
+					     region[i].addr,
+					     region[i].len)) {
+			EMSG("RISAF %#"PRIxPA": Regions %u and %u overlap",
+			     risaf->pdata.base.pa, index, i);
+			return TEE_ERROR_GENERIC;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
+					 uint32_t region_id, uint32_t cfg,
+					 uint32_t cid_cfg, paddr_t saddr,
+					 paddr_t eaddr)
+{
+	uint32_t mask = risaf->ddata->mask_regions;
+	vaddr_t base = risaf_base(risaf);
+
+	io_clrbits32(base + _RISAF_REG_CFGR(region_id), _RISAF_REG_CFGR_BREN);
+
+	io_clrsetbits32(base + _RISAF_REG_STARTR(region_id), mask,
+			(saddr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_ENDR(region_id), mask,
+			(eaddr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_CIDCFGR(region_id),
+			_RISAF_REG_CIDCFGR_ALL_MASK,
+			cid_cfg & _RISAF_REG_CIDCFGR_ALL_MASK);
+
+	io_clrsetbits32(base + _RISAF_REG_CFGR(region_id),
+			_RISAF_REG_CFGR_ALL_MASK,
+			cfg & _RISAF_REG_CFGR_ALL_MASK);
+
+	if (cfg & _RISAF_REG_CFGR_ENC) {
+		if (!risaf->pdata.enc_supported) {
+			EMSG("RISAF %#"PRIxPA": encryption feature error",
+			     risaf->pdata.base.pa);
+			return TEE_ERROR_GENERIC;
+		}
+
+		if ((cfg & _RISAF_REG_CFGR_SEC) != _RISAF_REG_CFGR_SEC) {
+			EMSG("RISAF %#"PRIxPA": encryption on non secure area",
```

Comment:
For consistency in OP-TEE OS source files, prefer "non-secure" over "non secure".

I wonder if it would be better to check these error cases before the RISAF config is changed.

Reply:
Sure, I'll move the checks above